man/dmlua.1
.TH "DMLUA" "1" "" "Version 2.0.0" "User Commands"
.SH NAME
dmlua \(en passes received observations to Lua function
.SH SYNOPSIS
\f[B]dmlua\f[R] \-\-help
.PP
\f[B]dmlua\f[R] \-\-version
.PP
\f[B]dmlua\f[R] \-\-\f[B]node\f[R] \f[I]id\f[R] \-\-\f[B]script\f[R]
\f[I]file\f[R] [\-\-\f[B]procedure\f[R] \f[I]name\f[R]]
[\-\-\f[B]name\f[R] \f[I]name\f[R]] [\-\-\f[B]logger\f[R]
\f[I]name\f[R]] [\-\-\f[B]debug\f[R]] [\-\-\f[B]verbose\f[R]]
.PP
\f[B]dmlua\f[R] \-\-\f[B]config\f[R] \f[I]file\f[R] [\-\-\f[B]name\f[R]
\f[I]name\f[R]]
.SH DESCRIPTION
The \f[B]dmlua\f[R] program runs a custom Lua script to process
observations received from POSIX message queue.
Each observation is passed as a Lua table to the Lua function of the
name given in option \f[CR]procedure\f[R].
The function name \f[CR]process\f[R] is assumed by default.
The function must return the (modified) observation table on exit.
.PP
The observation returned from the Lua function will be forwarded to the
next specified receiver or discarded if no receivers are left.
.PP
The Lua script file will be executed after loading.
Any code outside of the Lua function will be evaluated once to allow an
(optional) initialisation of the script.
.SH OPTIONS
.TP
\f[B]\-\-config\f[R], \f[B]\-c\f[R] \f[I]file\f[R]
File path to the optional configuration file.
.TP
\f[B]\-\-debug\f[R], \f[B]\-D\f[R]
Forward logs messages of level \f[CR]LL_DEBUG\f[R] via IPC (if logger is
set).
.TP
\f[B]\-\-help\f[R], \f[B]\-h\f[R]
Print available command\-line arguments and quit.
.TP
\f[B]\-\-logger\f[R], \f[B]\-l\f[R] \f[I]name\f[R]
Name of logger.
If set, sends logs to \f[I]dmlogger(1)\f[R] process of given name.
.TP
\f[B]\-\-name\f[R], \f[B]\-n\f[R] \f[I]name\f[R]
Name of program instance and configuration (default is
\f[CR]dmlua\f[R]).
.TP
\f[B]\-\-node\f[R], \f[B]\-N\f[R] \f[I]id\f[R]
Node id.
.TP
\f[B]\-\-procedure\f[R], \f[B]\-p\f[R] \f[I]name\f[R]
Lua function name (default is \f[CR]process\f[R]).
.TP
\f[B]\-\-script\f[R], \f[B]\-s\f[R] \f[I]file\f[R]
Path to Lua script file.
.TP
\f[B]\-\-verbose\f[R], \f[B]\-V\f[R]
Print log messages to \f[I]stderr\f[R].
.TP
\f[B]\-\-version\f[R], \f[B]\-v\f[R]
Print version information and quit.
.SH EXIT STATUS
.TP
\f[B]0\f[R]
Success.
Process terminated without errors.
.TP
\f[B]1\f[R]
Failure.
Process failed.
.SH ENVIRONMENT
.TP
\f[B]DM_LOGGER\f[R]
Name of logger instance to send logs to.
.TP
\f[B]NO_COLOR\f[R]
Disable ANSI colour output.
.SH EXAMPLE
Wait for an observation on POSIX message queue \f[CR]/dmlua\f[R], pass
it as Lua table \f[CR]observ\f[R] to function \f[CR]process(observ)\f[R]
in \f[CR]script.lua\f[R], then forward the result to the next receiver:
.IP
.EX
$ dmlua \-n dmlua \-N dummy\-node \-s script.lua
.EE


CHANGELOG.md
# Change Log

All notable changes to the project will be documented in this file.

## [2.0.0] – Unreleased

## Library

* `Added` error codes
* `Added` _fortran-nng_ interface bindings to NNG to `vendor/`
* `Added` module `dm_arg_parser` (outsourced from `dm_arg`)
* `Added` module `dm_group` for observation groups
* `Added` module `dm_ipc` for socket access
* `Added` module `dm_ipc_async` for asynchronous I/O
* `Added` module `dm_ipc_disco` for service discovery
* `Added` module `dm_ipc_message` for message handling
* `Added` module `dm_ipc_mutex` for mutex access
* `Added` module `dm_ipc_thread` for thread access
* `Added` module `dm_ipc_type` of IPC enumerators
* `Added` module `dm_job_list` (outsourced from `dm_job`)
* `Changed` names of modules `dm_fifo`, `dm_mqueue`, `dm_mqueue_util`, `dm_pipe`,
  `dm_sem`, `dm_signal`, `dm_thread`, `dm_tty` to include `posix` prefix
* `Changed` name of module `dm_system` to `dm_posix`
* `Changed` database schema due to new data models
* `Changed` default database suffix from `.sqlite` to `.db`
* `Changed` GeoCOM API for Lua to accept prototype observation as first argument
* `Changed` model of observation data structure `dm_observ`
* `Deleted` model of request data strucutre `dm_request`

## Programs

* `Changed` response value units in `dmved`

## Documentation

* Converted user guide from AsciiDoc to Markdown
* Converted man pages from AsciiDoc to Markdown

## [1.0.0] – Unreleased

## Library

* `Added` Bessel, Butterworth, Chebyshev filters to `dm_filter`

## [0.9.9] – 2026-01-11

### Library

* `Added` module `dm_coord` for coordinate transformation
* `Added` module `dm_filter` for filtering
* `Added` module `dm_ghostscript` for Ghostscript access
* `Added` module `dm_netstring` for netstring parsing
* `Added` module `dm_statistics` of statistics functions
* `Added` file tree size function to module `dm_file`
* `Added` file modification date/time to function `dm_file_touch()` in `dm_file`
* `Added` format `FORMAT_TSV` for tab-separated values to `dm_format`
* `Added` swap routines to `dm_util`
* `Changed` error code order for future additions
* `Changed` default database cache size of `dm_db_open()` to improve performance

### Programs

* `Added` author option to `dmreport`
* `Added` meta data to PDF output of `dmreport`
* `Fixed` reading of options in `dmdbctl`
* `Changed` last modified date/time to last log date/time in output of `dmfeed`

## [0.9.8] – 2025-08-30

### Library

* `Added` module `dm_image` for image handling
* `Added` module `dm_js` for JavaScript generation
* `Added` module `dm_roff` for GNU roff abstraction
* `Added` module `dm_serial` for derived type serialisation
* `Added` module `dm_transfer` for HTTP-RPC API file transfers
* `Added` GeoJSON Feature Collection procedures to `dm_geojson`
* `Added` image upload to `dm_rpc`
* `Added` HTTP request and response headers to `dm_rpc`
* `Added` HTTP response headers to `dm_fcgi`
* `Added` terminals `gpic`, `postscript`, and `sixeltek` to `dm_plot`
* `Added` image and transfer functions to `dm_db_api`
* `Changed` structure of database abstraction layer, added modules `dm_db_api`, `dm_db_count`, `dm_db_json`, `dm_db_pragma`, and `dm_db_row`
* `Changed` API of modules `dm_arg` and `dm_config` to object-oriented
* `Changed` query parameter API in `dm_cgi`
* `Changed` Git submodules to copies of third-party dependencies
* `Fixed` reading from pipe in `dm_pipe`
* `Fixed` unsigned type conversion in `dm_c`
* `Fixed` return code handling of database select functions

### Programs

* `Added` program `dmcamera` for capturing of camera images
* `Added` program `dmupload` for image upload to HTTP-RPC API
* `Added` scale factor for response values to `dmreport`
* `Added` PDF and PostScript output to `dmreport`
* `Added` message queue reading to `dmmb`
* `Added` image upload to `dmapi`
* `Added` image viewer to `dmweb`
* `Added` environment variables `DM_IMAGE_DB`, `DM_IMAGE_DIR` to `dmapi` and `dmweb`
* `Added` log file output to `dmlogger`
* `Changed` environment variables in `dmapi` and `dmweb` to `DM_BEAT_DB`, `DM_LOG_DB`, `DM_OBSERV_DB`
* `Changed` derived type serialisation in `dmapi` and `dmexport`
* `Changed` database access in `dmexport`
* `Changed` configuration of `dmgrc`

## [0.9.7] – 2025-04-05

### Library

* `Added` module `dm_c` for C interoperability
* `Added` module `dm_camera` for webcam and RTSP access (FFmpeg)
* `Added` module `dm_crypto` of cryptographic hash functions
* `Added` module `dm_db_query` for SQL query building
* `Added` module `dm_dwd` for DWD API format parsing
* `Added` module `dm_dwd_api` for DWD API access (libcurl)
* `Added` module `dm_ftp` for FTP transfer (libcurl)
* `Added` module `dm_freebsd` for system status access on FreeBSD
* `Added` module `dm_geojson` for GeoJSON serialisation
* `Added` module `dm_gm` for GraphicsMagick access
* `Added` module `dm_im` for XMPP connectivity (libstrophe)
* `Added` module `dm_linux` for system status access on Linux
* `Added` module `dm_modbus` for Modbus RTU/TCP connectivity (libmodbus)
* `Added` module `dm_ve` for Victron Energy VE.Direct protocol support
* `Added` library build date and system type parameters (CPP, FPP)
* `Added` Linux and FreeBSD abstractions in module `dm_system`
* `Changed` refactored database abstraction layer
* `Changed` refactored database schema
* `Changed` refactored command-line argument parsing
* `Changed` refactored logical function names
* `Fixed` wrong baud rate in GeoCOM class

### Programs

* `Added` program `dmbot` for remote control of clients through XMPP
* `Added` program `dmdwd` for Deutscher Wetterdienst (DWD) API monitoring
* `Added` program `dmmb` for sensor monitoring through Modbus RTU/TCP
* `Added` program `dmmbctl` for Modbus RTU/TCP command-line control
* `Added` program `dmsystem` for system monitoring
* `Added` program `dmved` for MPPT and battery monitoring (VE.Direct)
* `Added` interactive map view to `dmweb` (Leaflet)

### Documentation


app/dmlua.f90
! dmlua.f90
!
! Author:  Philipp Engel
! Licence: ISC
program dmlua
    !! Lua script spawner that reads an observation from POSIX message queue,
    !! passes it to a configured Lua script, and forwards the returned
    !! observation to the next specified receiver.
    use :: dmpack
    implicit none (type, external)

    character(*), parameter :: APP_NAME  = 'dmlua'
    integer,      parameter :: APP_MAJOR = 2
    integer,      parameter :: APP_MINOR = 0
    integer,      parameter :: APP_PATCH = 0

    integer, parameter :: APP_PROCEDURE_LEN = 32     !! Max. length of Lua function name.
    logical, parameter :: APP_MQ_BLOCKING   = .true. !! Observation forwarding is blocking.

    type :: app_type
        !! Global application settings.
        character(ID_LEN)            :: name      = APP_NAME  !! Instance and configuration name (required).
        character(FILE_PATH_LEN)     :: config    = ' '       !! Path to configuration file (required).
        character(LOGGER_NAME_LEN)   :: logger    = ' '       !! Name of logger.
        character(NODE_ID_LEN)       :: node_id   = ' '       !! Node id (required).
        character(APP_PROCEDURE_LEN) :: procedure = 'process' !! Name of Lua function (required).
        character(FILE_PATH_LEN)     :: script    = ' '       !! Path to Lua script file (required).
        logical                      :: debug     = .false.   !! Forward debug messages via IPC.
        logical                      :: verbose   = .false.   !! Print debug messages to stderr.
    end type app_type

    class(logger_class), pointer :: logger ! Logger object.

    integer                 :: rc     ! Return code.
    type(app_type)          :: app    ! App configuration.
    type(lua_state_type)    :: lua    ! Lua interpreter.
    type(posix_mqueue_type) :: mqueue ! Message queue.

    ! Initialise DMPACK.
    call dm_init()

    ! Get command-line arguments, read options from configuration file.
    rc = read_args(app)
    if (dm_is_error(rc)) call dm_stop(STOP_FAILURE)

    ! Initialise logger.
    logger => dm_logger_get_default()
    call logger%configure(name    = app%logger,  & ! Name of logger process.
                          node_id = app%node_id, & ! Node id.
                          source  = app%name,    & ! Log source.
                          debug   = app%debug,   & ! Forward debug messages via IPC.
                          ipc     = .true.,      & ! Enable IPC (if logger is set).
                          verbose = app%verbose)   ! Print logs to standard error.
    call logger%info('started ' // APP_NAME)

    call init(app, lua, mqueue, error=rc)
    if (dm_is_error(rc)) call shutdown(rc)

    call run(app, lua, mqueue)
    call shutdown(E_NONE)
contains
    subroutine shutdown(error)
        !! Cleans up and stops program.
        integer, intent(in) :: error

        integer :: rc, stat

        stat = merge(STOP_FAILURE, STOP_SUCCESS, dm_is_error(error))

        call dm_posix_mqueue_close(mqueue, error=rc)
        if (dm_is_error(rc)) call logger%error('failed to close mqueue /' // app%name, error=rc)

        call dm_posix_mqueue_unlink(mqueue, error=rc)
        if (dm_is_error(rc)) call logger%error('failed to unlink mqueue /' // app%name, error=rc)

        call dm_lua_destroy(lua)

        call logger%info('stopped ' // APP_NAME, error=error)
        call dm_stop(stat)
    end subroutine shutdown

    subroutine init(app, lua, mqueue, error)
        !! Initialises program.
        type(app_type),          intent(inout)         :: app    !! App type.
        type(lua_state_type),    intent(out)           :: lua    !! Lua state type.
        type(posix_mqueue_type), intent(out)           :: mqueue !! POSIX message queue type.
        integer,                 intent(out), optional :: error  !! Error code.

        integer :: rc

        init_block: block
            ! Initialise Lua interpreter.
            rc = dm_lua_init(lua)

            if (dm_is_error(rc)) then
                call logger%error('failed to init Lua interpreter', error=rc)
                exit init_block
            end if

            ! Register DMPACK API for Lua.
            rc = dm_lua_api_register(lua)

            if (dm_is_error(rc)) then
                call logger%error('failed to register Lua API', error=rc)
                exit init_block
            end if

            ! Register GeoCOM API for Lua.
            rc = dm_lua_geocom_register(lua, procedures=.false., errors=.true.)

            if (dm_is_error(rc)) then
                call logger%error('failed to register GeoCOM API', error=rc)
                exit init_block
            end if

            ! Open and run Lua script once.
            rc = dm_lua_open(lua, app%script, eval=.true.)

            if (dm_is_error(rc)) then
                call logger%error('failed to load Lua script ' // app%script, error=rc)
                exit init_block
            end if

            ! Open observation message queue for reading.
            rc = dm_posix_mqueue_open(mqueue, type=TYPE_OBSERV, name=app%name, access=POSIX_MQUEUE_RDONLY)

            if (dm_is_error(rc)) then
                call logger%error('failed to open mqueue /' // trim(app%name) // ': ' // dm_posix_error_message(), error=rc)
                exit init_block
            end if

            ! Register signal handlers.
            call dm_posix_signal_register(signal_callback)
        end block init_block

        if (present(error)) error = rc
    end subroutine init

    subroutine run(app, lua, mqueue)
        !! Waits for incoming observation, passes derived type as table to Lua
        !! function. The Lua function has to return the (modified) observation
        !! on exit.
        !!
        !! Variable `observ_in` stores the observation type received from message
        !! queue. The observation data returned from the Lua function is stored
        !! in `observ_out` and will be forwarded to the next receiver. On error,
        !! the received observation will be forwarded instead.
        type(app_type),          intent(inout) :: app
        type(lua_state_type),    intent(inout) :: lua
        type(posix_mqueue_type), intent(inout) :: mqueue

        integer           :: rc
        type(observ_type) :: observ_in, observ_out

        ipc_loop: do
            ! Blocking read from POSIX message queue.
            call logger%debug('waiting for observation on mqueue /' // app%name)
            rc = dm_posix_mqueue_read(mqueue, observ_in)

            if (dm_is_error(rc)) then
                call logger%error('failed to read observation from mqueue /' // app%name, error=rc)
                call dm_sleep(1)
                cycle ipc_loop
            end if

            ! Validate observation.
            if (.not. dm_observ_is_valid(observ_in)) then
                call logger%error('received invalid observation ' // trim(observ_in%name), observ=observ_in, error=E_INVALID)
                cycle ipc_loop
            end if

            call logger%debug('passing observation ' // trim(observ_in%name) // ' to Lua function ' // trim(app%procedure) // '()', observ=observ_in)

            ! Pass the observation to the Lua function and read the returned observation.
            lua_block: block
                ! Load Lua function.
                rc = dm_lua_read(lua, trim(app%procedure))

                if (dm_is_error(rc)) then
                    call logger%error('failed to load Lua function ' // trim(app%procedure) // '()', error=rc)
                    exit lua_block
                end if

                if (.not. dm_lua_is_function(lua)) then
                    rc = E_INVALID
                    call logger%error('invalid Lua function ' // trim(app%procedure) // '()', error=rc)
                    exit lua_block
                end if

                ! Write derived type to Lua stack and call the Lua function.
                call dm_lua_from(lua, observ_in)
                rc = dm_lua_call(lua, nargs=1, nresults=1)

                if (dm_is_error(rc)) then
                    call dm_lua_pop(lua)
                    call logger%error('failed to execute Lua function ' // trim(app%procedure) // '()', observ=observ_in, error=rc)
                    exit lua_block
                end if

                ! Read observation from Lua stack into derived type.
                rc = dm_lua_to(lua, observ_out)

                if (dm_is_error(rc)) then
                    call dm_lua_pop(lua)
                    call logger%error('failed to read observation from Lua stack', error=rc, observ=observ_in)
                    exit lua_block
                end if

                ! Validate returned observation.
                if (.not. dm_observ_is_valid(observ_out)) then
                    rc = E_INVALID
                    call logger%error('invalid observation returned from Lua function ' // trim(app%procedure) // '()', error=rc, observ=observ_in)
                    exit lua_block
                end if
            end block lua_block

            ! Forward observation. On error, send the original observation instead.
            if (dm_is_error(rc)) then
                call logger%debug('forwarding observation ' // trim(observ_in%name) // ' unmodified', observ=observ_out)
                rc = dm_posix_mqueue_forward(observ_in, name=app%name, blocking=APP_MQ_BLOCKING)
            else
                call logger%debug('forwarding observation ' // observ_out%name, observ=observ_out)
                rc = dm_posix_mqueue_forward(observ_out, name=app%name, blocking=APP_MQ_BLOCKING)
            end if
        end do ipc_loop
    end subroutine run

    ! **************************************************************************
    ! COMMAND-LINE ARGUMENTS AND CONFIGURATION FILE.
    ! **************************************************************************
    integer function read_args(app) result(rc)
        !! Reads command-line arguments and settings from configuration file.
        type(app_type), intent(out) :: app

        type(arg_parser_class) :: parser

        call parser%add('name',      short='n', type=ARG_TYPE_ID)                 ! -n, --name <id>
        call parser%add('config',    short='c', type=ARG_TYPE_FILE)               ! -c, --config <path>
        call parser%add('logger',    short='l', type=ARG_TYPE_ID)                 ! -l, --logger <id>
        call parser%add('node',      short='N', type=ARG_TYPE_ID)                 ! -N, --node <id>
        call parser%add('procedure', short='p', type=ARG_TYPE_STRING)             ! -p, --procedure <string>
        call parser%add('script',    short='s', type=ARG_TYPE_FILE, exist=.true.) ! -s, --script <path>
        call parser%add('debug',     short='D', type=ARG_TYPE_LOGICAL)            ! -D, --debug
        call parser%add('verbose',   short='V', type=ARG_TYPE_LOGICAL)            ! -V, --verbose

        ! Read all command-line arguments.
        rc = parser%read(version_callback)
        if (dm_is_error(rc)) return

        call parser%get('name',   app%name)
        call parser%get('config', app%config)

        ! Read configuration from file.
        rc = read_config(app)
        if (dm_is_error(rc)) return

        ! Get all other arguments.
        call parser%get('logger',    app%logger)
        call parser%get('node',      app%node_id)
        call parser%get('procedure', app%procedure)
        call parser%get('script',    app%script)
        call parser%get('debug',     app%debug)
        call parser%get('verbose',   app%verbose)

        ! Validate options.
        rc = validate(app)
    end function read_args

    integer function read_config(app) result(rc)
        !! Reads configuration from (Lua) file.
        type(app_type), intent(inout) :: app !! App type.

        type(config_class) :: config

        rc = E_NONE
        if (.not. dm_string_has(app%config)) return

        rc = config%open(app%config, app%name)

        if (dm_is_ok(rc)) then
            call config%get('logger',    app%logger)
            call config%get('node',      app%node_id)
            call config%get('procedure', app%procedure)
            call config%get('script',    app%script)
            call config%get('debug',     app%debug)
            call config%get('verbose',   app%verbose)
        end if

        call config%close()
    end function read_config

    integer function validate(app) result(rc)
        !! Validates options and prints error messages.
        character(*), parameter :: PROCEDURE_SET = &
            '-0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ_abcdefghijklmnopqrstuvwxyz' ! Valid procedure name characters.

        type(app_type), intent(inout) :: app !! App type.

        rc = E_INVALID

        if (.not. dm_id_is_valid(app%name)) then
            call dm_error_out(rc, 'invalid name')
            return
        end if

        if (dm_string_has(app%logger) .and. .not. dm_id_is_valid(app%logger)) then
            call dm_error_out(rc, 'invalid logger')
            return
        end if

        if (.not. dm_id_is_valid(app%node_id)) then
            call dm_error_out(rc, 'invalid node id')
            return
        end if

        if (verify(trim(app%procedure), PROCEDURE_SET) > 0) then
            call dm_error_out(rc, 'invalid Lua function name ' // app%procedure)
            return
        end if

        rc = E_NOT_FOUND

        if (.not. dm_file_exists(app%script)) then
            call dm_error_out(rc, 'Lua script ' // trim(app%script) // ' not found')
            return
        end if

        rc = E_NONE
    end function validate

    ! **************************************************************************
    ! CALLBACKS.
    ! **************************************************************************
    subroutine signal_callback(signum) bind(c)
        !! Default POSIX signal handler of the program.
        integer(c_int), intent(in), value :: signum

        call logger%debug('exit on on signal ' // dm_posix_signal_name(signum))
        call shutdown(E_NONE)
    end subroutine signal_callback

    subroutine version_callback()
        call dm_version_out(APP_NAME, APP_MAJOR, APP_MINOR, APP_PATCH)
        print '(a)', dm_lua_version(.true.)
    end subroutine version_callback
end program dmlua
